include/ck_tile/host/host_tensor.hpp

```diff
@@ -271,253 +291,597 @@ CK_TILE_HOST auto make_ParallelTensorFunctor(F f, Xs... xs)
     {
+        assert(mView.get_num_of_dimension() <= MaxNumDims);
+        assert(std::size(repeats) <= mView.get_num_of_dimension());
+
+        using std::rbegin, std::rend;
+        std::copy(rbegin(repeats), rend(repeats), rbegin(get_repeats()));
     }
 
-    HostTensor(const Descriptor& desc) : mDesc(desc), mData(mDesc.get_element_space_size()) {}
+    size_type get_num_of_dimension() const { return mView.get_num_of_dimension(); }
 
-    template <typename OutT>
-    HostTensor<OutT> CopyAsType() const
+    size_type get_length(size_type dim) const { return mView.get_length(dim) * get_repeat(dim); }
+
+    auto get_lengths() const
     {
-        HostTensor<OutT> ret(mDesc);
-        std::transform(mData.cbegin(), mData.cend(), ret.mData.begin(), [](auto value) {
-            return ck_tile::type_convert<OutT>(value);
-        });
-        return ret;
+        return make_transform_range(make_zip_range(mView.get_lengths(), get_repeats()),
+                                    [](auto bundle) {
+                                        // length * repeat
+                                        return std::get<0>(bundle) * std::get<1>(bundle);
+                                    });
     }
 
-    HostTensor()                  = delete;
-    HostTensor(const HostTensor&) = default;
-    HostTensor(HostTensor&&)      = default;
+    template <typename... Is>
+    std::enable_if_t<((std::is_integral_v<Is> && std::is_convertible_v<Is, size_type>)&&...),
+                     const_reference>
+    operator()(Is... is) const
+    {
+        return (*this)(std::array{static_cast<size_type>(is)...});
+    }
 
-    ~HostTensor() = default;
+    const_reference operator()(span<const size_type> idx) const
+    {
+        assert(std::size(idx) == get_num_of_dimension());
 
-    HostTensor& operator=(const HostTensor&) = default;
-    HostTensor& operator=(HostTensor&&) = default;
+        std::array<size_type, MaxNumDims> real_idx;
+        for(size_type dim = 0; dim < std::size(idx); ++dim)
+        {
+            real_idx[dim] = idx[dim] / get_repeat(dim);
+        }
 
-    template <typename FromT>
-    explicit HostTensor(const HostTensor<FromT>& other) : HostTensor(other.template CopyAsType<T>())
+        return mView(span<const size_type>(std::data(real_idx), std::size(idx)));
+    }
+
+    private:
+    size_type get_repeat(size_type dim) const { return mRepeats[dim]; }
+
+    auto get_repeats()
+    {
+        using std::begin, std::next;
+
+        return iterator_range(begin(mRepeats), next(begin(mRepeats), get_num_of_dimension()));
+    }
+
+    auto get_repeats() const
+    {
+        using std::begin, std::next;
+
+        return iterator_range(begin(mRepeats), next(begin(mRepeats), get_num_of_dimension()));
+    }
+
+    TensorView mView;
+    std::array<size_type, MaxNumDims> mRepeats;
+};
+} // namespace detail
+
+template <typename T>
+struct HostTensorView : private HostTensorDescriptor
```

Comment:
HostTensor.hpp contain many useful data structure now.
Can we move implementation into HostTensor.cpp? Just like what old  CK does.
It is much easier to understand API/operation of these stuff.

Reply:
Because it's a class template. We cannot hide template implementation form the caller.